spark/src/main/spark-3.3/org/apache/comet/parquet/CometParquetFileFormat.scala

```diff
@@ -0,0 +1,231 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.comet.parquet
+
+import scala.collection.JavaConverters
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.parquet.filter2.predicate.FilterApi
+import org.apache.parquet.hadoop.ParquetInputFormat
+import org.apache.parquet.hadoop.metadata.FileMetaData
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.util.CaseInsensitiveMap
+import org.apache.spark.sql.catalyst.util.RebaseDateTime.RebaseSpec
+import org.apache.spark.sql.execution.datasources.DataSourceUtils
+import org.apache.spark.sql.execution.datasources.PartitionedFile
+import org.apache.spark.sql.execution.datasources.RecordReaderIterator
+import org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat
+import org.apache.spark.sql.execution.datasources.parquet.ParquetOptions
+import org.apache.spark.sql.execution.datasources.parquet.ParquetReadSupport
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.internal.SQLConf.LegacyBehaviorPolicy
+import org.apache.spark.sql.sources.Filter
+import org.apache.spark.sql.types.{DateType, StructType, TimestampType}
+import org.apache.spark.util.SerializableConfiguration
+
+import org.apache.comet.CometConf
+import org.apache.comet.MetricsSupport
+import org.apache.comet.shims.ShimSQLConf
+import org.apache.comet.vector.CometVector
+
+/**
+ * A Comet specific Parquet format. This mostly reuse the functionalities from Spark's
+ * [[ParquetFileFormat]], but overrides:
+ *
+ *   - `vectorTypes`, so Spark allocates [[CometVector]] instead of it's own on-heap or off-heap
+ *     column vector in the whole-stage codegen path.
+ *   - `supportBatch`, which simply returns true since data types should have already been checked
+ *     in [[org.apache.comet.CometSparkSessionExtensions]]
+ *   - `buildReaderWithPartitionValues`, so Spark calls Comet's Parquet reader to read values.
+ */
+class CometParquetFileFormat extends ParquetFileFormat with MetricsSupport with ShimSQLConf {
```

Comment:
The 3.2, 3.3, and 3.4 versions of `CometParquetFileFormat` are identical copies, and the 3.5 version only differs by one import.

I don't think we yet have a mechanism to share one class across multiple shim versions, so I don't think this is a blocker for this PR, but we probably want to look at implementing a way to avoid having large copies of code in the shim layer. I will file an issue for this. Maybe @kazuyukitanimura has thoughts on this too.

Reply:
I reviewed all of the shimmed classes and see that there are just two differences between 3.2-3.4 and 3.5.

The first issue is that `LegacyBehaviorPolicy` moved and this affects `CometParquetFileFormat`, `CometParquetPartitionReaderFactory`, and `ParquetFilters`.

```
< import org.apache.spark.sql.internal.SQLConf.LegacyBehaviorPolicy
---
> import org.apache.spark.sql.internal.LegacyBehaviorPolicy
```

The second issue is this change that affects `CometScanExec`.

```
274c274
<             PartitionedFileUtil.getPartitionedFile(f, f.getPath, p.values)
---
>             PartitionedFileUtil.getPartitionedFile(f, /*f.getPath,*/ p.values)
364c364
<               filePath = filePath,
---
>               //filePath = filePath,
```

I think that we could potentially avoid copying all of these classes and have more fine-grained shim changes. I will look into this later today and make a proposal.

Reply:
I now see that we do have a `spark-3.x` shim folder. We could also add a `spark-pre-3.5` folder to reduce the number of classes that are duplicated in this PR.

Reply:
I second that we should minimize copying as much as possible. Also, perhaps rebasing on the `main` branch first may help

Reply:
How was this done for Spark 4.0?